### ISSUE TITLE
The quest for better tessellation

### ISSUE THREAD

**Issue by ziyaointl** [p5py/p5#173]
This issue attempts to explore some of the options for

1. Improving tessellation/triangulation performance
2. Support tessellation/triangulation in 3D

### Current implementation in p5py
The existing implementation uses the `triangle` python module to triangulate shapes before sending them to the GPU. The `triangle` module is in turn a wrapper around the [Triangle](http://www.cs.cmu.edu/~quake/triangle.html) C library written by [Jonathan Shewchuck](https://people.eecs.berkeley.edu/~jrs/) that implements the Delaunay triangulation algorithm.

This [site](http://vterrain.org/Implementation/Libs/triangulate.html) notes that the Triangle Library by Prof. Shewchuck has several deficiencies. In particular,
> 1. It does not like duplicate vertices or duplicate edges.  'Duplicate' in this case is relative to numeric precision: For a building 10-100 meters in size, two vertices within 8cm of each other, defining a very short edge, can cause Triangle to crash.
> 2. It does not like it when a hole (inner ring) in the polygon has a vertex in the same location as one in the outer ring (crash).

In fact, the problems mentioned above are rather minor. The biggest problem of all is probably the fact that `triangle` does not support 3D points.

### Current implementation in Processing and p5.js
Maybe we can reference the triangulation implementations in other branches of the processing language? Here's what I found:

For Processing (OpenGL backend), GLU was referenced through the JOGL binding, evidenced by source files [PGraphicsOpenGL.java#L10985](https://github.com/processing/processing/blob/4bb41d9851f24584c064d75e759e6ac3b0f65928/core/src/processing/opengl/PGraphicsOpenGL.java#L10985) and [PJOGL.java#L605](https://github.com/processing/processing/blob/4bb41d9851f24584c064d75e759e6ac3b0f65928/core/src/processing/opengl/PJOGL.java#L605). Given that we don't have JOGL on Python and GLFW doesn't support GLU, we might have to look for other options.

For p5.js, tessellation support was achieved via [libtess.js](https://github.com/brendankenny/libtess.js/), which uses the GLU tessellation algorithm but rewritten in JavaScript ([reference](https://github.com/processing/p5.js/blob/730f75738b0b7e0443340a6e7daf928402c47471/contributor_docs/project_wrapups/adilrabbani_gsoc_2018.md)). 

#### OpenGL/GLU
From first glance, using OpenGL is as simple as using the following API:
```c
gluTessBeginPolygon(tess, user_data);
    gluTessBeginContour(tess);
        gluTessVertex(tess, coords, vertex_data);
        ...
        gluTessVertex(tess, coords_n, vertex_data);
    gluTessEndContour(tess);
gluTessEndPolygon(tess);
```
(Code example adapted from [songho.ca](http://www.songho.ca/opengl/gl_tessellation.html))

However, further searches indicated that this API is not part of OpenGL core, but part of GLU (OpenGL Utility Library). According to [GLFW](https://www.glfw.org/docs/3.3/build_guide.html)
> GLU has been deprecated and should not be used in new code.

So GLFW isn't helpful here.

I asked the question of how to use GLU bindings on the vispy Gitter chat after not finding anything related to GLU on their documentation. I did get a response from vispy maintainers very quickly. Thank you! However, he confirmed that vispy does not currently support the tessellation API.

>  Short answer is no. Long answer: I'm not familiar with the gluTess API specifically, but I'm guessing this has to do with tessellation shaders, right? Currently VisPy does not have any support for tessellation shaders but I think there are PRs trying to add it. In the last year we added support for geometry shaders which really set up the interfaces for adding additional shaders like the tessellation shaders, but so far these have not been added.
>  ---@djhoese

Given that our current dependencies don't have this functionality, another option is to introduce a new library called   [PyOpenGL](http://pyopengl.sourceforge.net/documentation/manual-3.0/gluTessBeginPolygon.html), which offers a wrapper for GLU (in addition to a lot more OpenGL functionality). Since we already have `vispy.gloo` for making OpenGL calls, we may want to discuss whether it's worth it to introduce another dependency with potentially the same capability as `vispy` but only use it for tessellation.

#### Scipy
[Scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.Delaunay.html) also has a function that performs the Delaunay triangulation algorithm. Instead of wrapping around Prof. Shewchuck's library, it wraps around [Qhull](http://www.qhull.org/). This supports both 2D and 3D. However, the Qhull website [says]() the library only works with convex hulls, so non-convex objects (e.g. objects with holes created by `begin_contour`) wouldn't work.

### Moving Forward
Between scipy and PyOpenGL, I am leaning towards PyOpenGL because it uses the same underlying algorithm as the Processing Language and p5.js while also supporting non-convex shapes. The downside is that it's a rather large library that overlaps in features with vispy. If you have an opinion on this subject, leave a comment down below!

### Roadmap
Here's my tentative plan to improve tessellation in p5py:

- [x] Build sample scenes for profiling
- [x] Profile performance of master branch as baseline
- [x] Fix obvious performance issues (if any)
- [x] Replace `triangle` with PyOpenGL, or another triangulation library
- [x] Profile again and compare to baseline
- [x] Implement parts of `begin_shape` and `end_shape` that only requires trivial/no triangulation in 3D (i.e. `POINTS`, `LINES`, `TRIANGLES`, `TRIANGLE_FAN`, `TRIANGLE_STRIP`, `QUADS`, and `QUAD_STRIP`).
- [x] Implement the complete version of `begin_shape` and `end_shape` in 3D (i.e. `TESS` in addition to the modes mentioned above).
- [x] (Stretch goal) Implement `begin_contour` and `end_contour` in 3D.
- [x] Document API changes

**Comment by parsoyaarihant** [comment]
> It does not like duplicate vertices or duplicate edges. 'Duplicate' in this case is relative to numeric precision: For a building 10-100 meters in size, two vertices within 8cm of each other, defining a very short edge, can cause Triangle to crash.

I have experienced this while implementing `begin_shape()`. As a workaround p5py checks for duplicate vertices to prevent this crash: https://github.com/p5py/p5/blob/master/p5/core/shape.py#L239

> For p5.js, tessellation support was achieved via libtess.js, which uses the GLU tessellation algorithm but rewritten in JavaScript (reference).

From what I know, p5.js doesn't currently tesselate the shapes in 2D. It directly uses the HTML canvas API: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D 

#167 was a recent bug in triangulation. While trying out PyOpenGL, you could try to tesselate this shape and check if its giving the intended output.

**Comment by parsoyaarihant** [comment]
I wanted some clarifications regarding the begin_shape() functionality in 3D.

When using begin_shape() in 3D, and all the vertices are not in the same plane, we would have to use tessellation in 3D right? It would be a 3D surface (or volumetric) tessellation rather than the generic 2D tessellation which we are currently using in p5py.

**Comment by ziyaointl** [comment]
> > It does not like duplicate vertices or duplicate edges. 'Duplicate' in this case is relative to numeric precision: For a building 10-100 meters in size, two vertices within 8cm of each other, defining a very short edge, can cause Triangle to crash.
> 
> I have experienced this while implementing `begin_shape()`. As a workaround p5py checks for duplicate vertices to prevent this crash: https://github.com/p5py/p5/blob/master/p5/core/shape.py#L239

Interesting, so it isn't an obscure edge case but a relevant one.

> > For p5.js, tessellation support was achieved via libtess.js, which uses the GLU tessellation algorithm but rewritten in JavaScript (reference).
> 
> From what I know, p5.js doesn't currently tesselate the shapes in 2D. It directly uses the HTML canvas API: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D

I think you are right - the default 2D renderer in p5.js does use the HTML canvas API. However, there is an [option](https://github.com/processing/p5.js/blob/7449231c082681ae69e794e93a63e1615239a991/src/webgl/p5.RendererGL.Immediate.js#L169) to tessellate user-defined shapes when using WebGL.
Another interesting finding is that it seems the Processing Language's 2D mode uses Path2D in JavaFX to [tessellate shapes](https://github.com/processing/processing/blob/0ac37bca1d0c7c12667a9c0628fba63b31c1f135/core/src/processing/javafx/PGraphicsFX2D.java#L338).

> #167 was a recent bug in triangulation. While trying out PyOpenGL, you could try to tesselate this shape and check if its giving the intended output.

Good point. I'll look into it once PyOpenGL is in place.

> I wanted some clarifications regarding the begin_shape() functionality in 3D.
> When using begin_shape() in 3D, and all the vertices are not in the same plane, we would have to use tessellation in 3D right? It would be a 3D surface (or volumetric) tessellation rather than the generic 2D tessellation which we are currently using in p5py.

This is a good question to ask. Initially, I assumed that p5.js is doing some fancy 3D tessellation. Turns out, GLUTess is only well-behaved when the points are co-planar, much like the `triangle` library that p5py is using ([reference](https://github.com/processing/p5.js/pull/3680#issuecomment-480999192)). This means the switch to GLUTess is now mostly a performance concern, aside from possibly fixing #167.

**Comment by ziyaointl** [comment]
Some ideas for new modes that we can add to `begin_shape` in 3D:

1. Project the points to 2D, tessellate them, and restore their z value. (Suggested [here](https://stackoverflow.com/a/52210455/9508687))
2. Assume the vertices define a volume and draw out their convex hull using [scipy.spatial.ConvexHull](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.ConvexHull.html)

**Comment by parsoyaarihant** [comment]
> Project the points to 2D, tessellate them, and restore their z value. (Suggested here)

A workaround this can be to transform the coordinate system to the plane in which the user wants to render the points. Then use begin_shape() API. So we dont have to detect the plane from the 3D points prior to tessellation. 

> Assume the vertices define a volume and draw out their convex hull using scipy.spatial.ConvexHull

This could be really cool feature for users working with point cloud data in 3D: https://stackoverflow.com/questions/27270477/3d-convex-hull-from-point-cloud

Additionally, we can add another API like `begin_volume()` for this functionality.

**Comment by ziyaointl** [comment]
> A workaround this can be to transform the coordinate system to the plane in which the user wants to render the points. Then use begin_shape() API. So we dont have to detect the plane from the 3D points prior to tessellation.

A naive approach (the one that I had in mind) was to always use the plane z = 0 for the projection and let the user transform the points however they see fit. I am not sure if I understood the "plane in which t he user wants to render the points" part. Could you elaborate?

> This could be really cool feature for users working with point cloud data in 3D.
> Additionally, we can add another API like begin_volume() for this functionality.

I agree! 

One suggestion: `begin_volume` might imply that we are doing some form of volumetric rendering (which we could, eventually, but may be slightly out of the scope for this summer). I think adding a mode like `HULL` or `CONVEX_HULL` to `begin_shape` might be a better idea.

**Comment by djhoese** [comment]
Hi! A quick note about pyopengl and vispy: Vispy has a light dependency on pyopengl for some features and often we have users install pyopengl. If there is some utility or interface that vispy can add that would make this easier for you I'd be willing to discuss it.

**Comment by ziyaointl** [comment]
> Hi! A quick note about pyopengl and vispy: Vispy has a light dependency on pyopengl for some features and often we have users install pyopengl. If there is some utility or interface that vispy can add that would make this easier for you I'd be willing to discuss it.

We haven't decided specifically how PyOpenGL might be used yet but as I attempt to integrate it with p5py we should have a better idea of whether any changes on the vispy side would be beneficial. Thank you so much for offering to help!

**Comment by ziyaointl** [comment]
I wrote two scenes for profiling: `basic_shapes` and `custom_shapes`. The `basic_shapes` scene was adpated from the `PShape` tutorial.

![Screenshot from 2020-06-07 14-38-14](https://user-images.githubusercontent.com/18119047/83980713-b9124b00-a8cc-11ea-9d7e-d297081c6410.png)

The `custom_shapes` scene creates 10 new 'C' shapes using `begin_shape` every frame and draws them on the screen.

![Screenshot from 2020-06-07 14-35-46](https://user-images.githubusercontent.com/18119047/83980718-c2031c80-a8cc-11ea-8aba-4bf411c4dffc.png)

Since we are mainly looking at triagulation, here are some interesting results from running `custom_shapes` for 1000 frames.

![Screenshot from 2020-06-07 14-41-27](https://user-images.githubusercontent.com/18119047/83980761-11494d00-a8cd-11ea-9f17-a1caa885df43.png)


1. Despite triangulating every frame, the total time spent in `_retriangulate` and `triangulate` is not particularly high compared to the total time spent in `on_timer`, only 1.272s and 0.7882s (not shown in the above chart).

2. `render_line` takes a total of 18.6s to execute, which is more than half of the total time spent in `on_timer`.

I don't observe any low-hanging fruit for performance optimization, so I'll proceed to look into using PyOpenGl and possibly fixing #167. If you want to take a look at the profiling data / run the code yourself, you can find them on the tessellation branch in the profiling folder. The `.prof` files were generated using cProfile and I used snakeviz to navigate them.

**Comment by parsoyaarihant** [comment]
Interesting. `render_line` does lot of preprocessing before sending the vertices to the shaders. 

I have run the profiling myself. I think there are a whole range of issues leading to poor performance. From what I observed, we are losing time performing unnecessary (or not very important) computations:

1. PShape does additional computations for more functionality. For example,  `_sanitize_vertex_list` is called every-time the PShape object is declared. `_transform_matrix` of every shape is initiated when PShape is declared. As far as I know, `_transform_matrix` functionality is only used while parsing SVG (converting SVG into PShapes). Additionally, every shape has its own stroke_weight, stroke_cap, and a whole range of properties.

2. In Color class, I am not sure how important it is to call `_recompute_hsb` every-time Color is called or any of its property is changed. 

3. In Renderer2D, we are converting the arrays into np arrays multiple times and changing their datatypes to np.uint32.

**Comment by parsoyaarihant** [comment]
@ziyaointl , I think the bigger goal is to improve the overall performance of p5py. Fixing #167 does not necessarily improve any performance. However, if you have time I would encourage you to work on #167 but keep it low in priority.

**Comment by ziyaointl** [comment]
Progress update: This image was tessellated using PyOpenGL!
![Screenshot from 2020-06-14 02-36-26](https://user-images.githubusercontent.com/18119047/84590033-ea68aa80-ade7-11ea-8ee3-a072cd57e7a0.png)

Here are some detailed todos:
- [ ] Support borders: either through sequential line drawing or through another instance of GluTess with `GLU_TESS_BOUNDARY_ONLY` set to true.
- [ ] Add contour support using PyOpenGL
- [ ] Refactor pipeline to not use triangle
- [ ] Profile performance difference

**Comment by ziyaointl** [comment]
Another update:
All GluTess stuff are pretty much in place. However, they are only hacked into the rendering pipeline and seem very out of place (i.e. they write to a custom draw queue that takes precedence over the standard vertex/edge properties). If you are curious of what I am talking about, feel free to take a look at the `tessllation` branch. I labeled all new variables and some new methods with the prefix `temp`.

To get GluTess fully integrated, I think I'll need to do a moderate amount of refactoring of the current rendering pipeline. As a bonus, this refactoring is likely going to introduce many performance improvements. For example, `TRIANGLE_STRIP` mode is noticably faster on my Intel HD620 because instead of drawing individual triangles (and making a draw call for each) I was able to combine the meshes and the lines into 2 draw calls for each pair of `begin_shape` and `end_shape`.

**Comment by parsoyaarihant** [comment]
I had a look at `tesselation`. It looks good. 

From the profiling code you have written. How much improvements have seen so far?

Does your approach supports all the shapes? (I think it does). Then it makes sense to remove the old functions.

**Comment by ziyaointl** [comment]
> I had a look at `tesselation`. It looks good.

Thanks for looking through my draft! I really appreciate it.

> From the profiling code you have written. How much improvements are you getting so far?

I have not ran the profiling code yet, only tests to check my implementation's correctness. I shall do some profiling next week.

> Does your approach supports all the shapes? (I think it does). Then it makes sense to remove the old functions to make the code more coherent.

Yep, the design does support all the shapes. However, there is still some work to be done to fully integrate it with existing code (e.g. passing `enum SType` instead of `attribs` in shape creation). Getting this done will be my priority for the short term.
